### PR TITLE
1

### PR DESCRIPTION
**Pull request by Las14**

### CODE REVIEW

File: text-analyzer-backend/src/main/java/textanalyzer/service/TextAnalyzerServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.Map;
+
+@Service
+public class TextAnalyzerServiceImpl implements TextAnalyzerService{
+
+
+    @Override
+    public Map<Character, Integer> analyzeText(Mode mode, String input) {
+        Map<Character, Integer> result = new HashMap<>();
+        char[] chars = input.toCharArray();
```
Comment (Las14) [inline text-analyzer-backend/src/main/java/textanalyzer/service/TextAnalyzerServiceImpl.java:16]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;input&#39; (index: 1 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;mode&#39;. The same parameter type is validated here for example: textAnalyzer6/text-analyzer-backend/src/main/java/textanalyzer/controller/TextAnalyzerController.java:34. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).

File: text-analyzer-backend/src/main/java/textanalyzer/controller/TextAnalyzerController.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    private static final Logger logger = LoggerFactory.getLogger(TextAnalyzerController.class);
+
+    private final TextAnalyzerService textAnalyzerService;
+
+    public TextAnalyzerController(TextAnalyzerService textAnalyzerService) {
+        this.textAnalyzerService = textAnalyzerService;
+    }
+
+    @PostMapping
+    public Map<Character, Integer> analyzeText(@RequestParam(required = true) String mode, @RequestParam(required = true) String text) {
```
Comment (Las14) [inline text-analyzer-backend/src/main/java/textanalyzer/controller/TextAnalyzerController.java:26]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Your code is using a vulnerable version of the Spring Framework that contains a remote code execution vulnerability exploitable through malicious HTTP requests. This poses a significant security risk as attackers can gain complete control of affected systems, potentially leading to data breaches, service disruption, or further network compromise. To remediate this issue, update to the latest patched version of Spring Framework (specific version depends on which CVE is affecting your installation) and consider implementing additional security layers such as web application firewalls that can detect and block malicious HTTP requests.

[Learn more](https://cwe.mitre.org/data/definitions/94.html)

File: text-analyzer-frontend/src/main.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
+
+import { AppModule } from './app/app.module';
+
+platformBrowserDynamic().bootstrapModule(AppModule, {
+  ngZoneEventCoalescing: true
+})
+  .catch(err => console.error(err));
```
Comment (Las14) [inline text-analyzer-frontend/src/main.ts:8]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Log Injection occurs when untrusted user input is directly written to log files without proper sanitization. This can allow attackers to manipulate log entries, potentially leading to security issues like log forging or cross-site scripting. To prevent this, always sanitize user input before logging by removing or encoding newline characters, using string encoding functions, and leveraging built-in sanitization features of logging libraries when available. Learn more - https://cwe.mitre.org/data/definitions/117.html

File: text-analyzer-backend/src/main/java/textanalyzer/service/TextAnalyzerServiceImpl.java (hunk 1 above)
Comment (Las14) [inline text-analyzer-backend/src/main/java/textanalyzer/service/TextAnalyzerServiceImpl.java:16]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;input&#39; (index: 1 | type: String). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other validated parameters: &#39;mode&#39;. The same parameter type is validated here for example: textAnalyzer6/text-analyzer-backend/src/main/java/textanalyzer/controller/TextAnalyzerController.java:34. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).

File: text-analyzer-backend/src/main/java/textanalyzer/controller/TextAnalyzerController.java (hunk 2 above)
Comment (Las14) [inline text-analyzer-backend/src/main/java/textanalyzer/controller/TextAnalyzerController.java:26]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Your code is using a vulnerable version of the Spring Framework that contains a remote code execution vulnerability exploitable through malicious HTTP requests. This poses a significant security risk as attackers can gain complete control of affected systems, potentially leading to data breaches, service disruption, or further network compromise. To remediate this issue, update to the latest patched version of Spring Framework (specific version depends on which CVE is affecting your installation) and consider implementing additional security layers such as web application firewalls that can detect and block malicious HTTP requests.

[Learn more](https://cwe.mitre.org/data/definitions/94.html)

File: text-analyzer-frontend/src/main.ts (hunk 3 above)
Comment (Las14) [inline text-analyzer-frontend/src/main.ts:8]: Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Log Injection occurs when untrusted user input is directly written to log files without proper sanitization. This can allow attackers to manipulate log entries, potentially leading to security issues like log forging or cross-site scripting. To prevent this, always sanitize user input before logging by removing or encoding newline characters, using string encoding functions, and leveraging built-in sanitization features of logging libraries when available. Learn more - https://cwe.mitre.org/data/definitions/117.html

File: text-analyzer-backend/src/main/java/textanalyzer/TextAnalyzerApplication.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+public class TextAnalyzerApplication {
+
+	public static void main(String[] args) {
+		SpringApplication.run(TextAnalyzerApplication.class, args);
+
+	}
+	public void isrunning(){
+
+	}
+
```
Comment (dev5-eng) [inline text-analyzer-backend/src/main/java/textanalyzer/TextAnalyzerApplication.java:16]: this function is not used and shall be removed

File: text-analyzer-backend/src/main/java/textanalyzer/controller/TextAnalyzerController.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+        Mode analysisMode;
+        int sum = 100;
```
Comment (dev5-eng) [inline text-analyzer-backend/src/main/java/textanalyzer/controller/TextAnalyzerController.java:29]: This variable is not used

File: text-analyzer-backend/src/test/java/text_analyzer_app/demo/TextAnalyzerTest.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+    @Test
+     void testAnalyzeTextConsonants() {
+        String input = "Text";
+        Map<Character, Integer> result = textAnalyzerService.analyzeText(Mode.CONSONANTS, input);
+
+        assertEquals(1, result.get('X'));
+        assertEquals(0, result.getOrDefault('E', 0));
+        assertEquals(0, result.getOrDefault('A', 0));
+    }
```
Comment (dev5-eng) [inline text-analyzer-backend/src/test/java/text_analyzer_app/demo/TextAnalyzerTest.java:42]: Consider more tests for edge cases and invalid and empty values

File: text-analyzer-frontend/src/app/app.component.html [hunk 7]
```diff
@@ -0,0 +1,44 @@
+<div class="container">
+  <div style="margin-bottom: 100px;">
+    <h1 style="color: rgb(12, 12, 194);" class="h1Text">Text Analyzer</h1>
+  </div>
+  <div class="input-container">
+  <mat-form-field appearance="fill" style="margin-bottom: 20px;"> 
+    <input matInput [(ngModel)]="inputText" class="inputText">
+  </mat-form-field>
+  
+  <mat-button-toggle-group name="analysisMode" [(ngModel)]="analysisMode" class="toggle-group">
```
Comment (dev5-eng) [inline text-analyzer-frontend/src/app/app.component.html:10]: Fix indentation here

File: text-analyzer-frontend/src/app/app.component.html [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+  <label>
+    <input type="checkbox" [checked]="onlineMode" (change)="toggleOnlineMode($event)" class="onlineMode"> Online Mode
+  </label>
+ 
+  <button mat-button (click)="analyzeText()" class="mat-button" style="color: white;">Analyze</button>
+
+  <button mat-button (click)="clearEntries()" class="mat-button" style="color: rgb(255, 0, 0);">Clear</button>
+</div>
+<div  *ngIf="analysisResult">
```
Comment (dev5-eng) [inline text-analyzer-frontend/src/app/app.component.html:25]: extra space here

File: text-analyzer-frontend/src/app/app.component.html [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+  <h2>Analysis Result</h2>
+  <div *ngFor="let entry of (analysisResult | keyvalue)">
+    Letter '{{ entry.key }}' appears {{ entry.value }} times
+  </div>
+</div>
+
+<div  *ngIf="previousResults.length > 0">
```
Comment (dev5-eng) [inline text-analyzer-frontend/src/app/app.component.html:32]: same extra space here before *ngIf

File: text-analyzer-frontend/src/app/app.component.html [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  <h2>Previous Results</h2>
+  <div *ngFor="let result of previousResults">
+    <div *ngFor="let entry of (result | keyvalue)">
+          Letter '{{ entry.key }}' appears {{ entry.value }} times
+</div>
```
Comment (dev5-eng) [inline text-analyzer-frontend/src/app/app.component.html:37]: Fix indentation here, consider also some automatic formatter like "prettier"

File: text-analyzer-backend/src/main/java/textanalyzer/service/TextAnalyzerServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+        return result;
+    }
+
+    private boolean isVowel(char c) {
+        return "AEIOUYaeiouy".indexOf(c) != -1;
+    }
+
+    private boolean isConsonant(char c) {
+        return true;
```
Comment (dev5-eng) [inline text-analyzer-backend/src/main/java/textanalyzer/service/TextAnalyzerServiceImpl.java:34]: This is wrong, function should contain some logic in it
